From d6f4f3b76f4069a7e0902a137cf4e4dd697b654d Mon Sep 17 00:00:00 2001
From: Binh Nguyen <binh.nguyen.uw@renesas.com>
Date: Mon, 25 Mar 2019 18:12:45 +0700
Subject: [PATCH] plat: renesas: bl2: add ECC support for DRAM

This commit reset all setting of ECC as default.
Also, all ECC registed must be initialized in IPL, if not,
U-boot or Kernel can not access to these registers.

User can use U-boot command to make ECC setting.

There is option build for different platforms, since they use
different register set:
- RZG_DRAM_EK874_ECC is for EK874 board

Signed-off-by: Binh Nguyen <binh.nguyen.uw@renesas.com>
---
 plat/renesas/rcar/bl2_fusa.c       | 64 ++++++++++++++++++++++++++++++++++++++
 plat/renesas/rcar/bl2_rcar_setup.c |  7 +++++
 plat/renesas/rcar/platform.mk      |  7 +++++
 3 files changed, 78 insertions(+)
 create mode 100644 plat/renesas/rcar/bl2_fusa.c

diff --git a/plat/renesas/rcar/bl2_fusa.c b/plat/renesas/rcar/bl2_fusa.c
new file mode 100644
index 0000000..290565a
--- /dev/null
+++ b/plat/renesas/rcar/bl2_fusa.c
@@ -0,0 +1,64 @@
+/*
+ * bl2_ecc_setup.c -
+ *
+ * Copyright (C) 2018 Renesas Electronics Corporation.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdint.h>
+#include <debug.h>
+#include <mmio.h>
+
+#define	DFUSAAREACR		0xE6785000	/* DRAM FuSa Area Conf */
+#define	DECCAREACR		0xE6785040	/* DRAM ECC Area Conf */
+#define	NUM_DAREA		16
+
+/* As the saddr, specify high-memory address (> 4 GB) */
+#define	FUSAAREACR(en, size, saddr)	\
+	(((uint32_t)en << 31) | ((uint32_t)size << 24) | (uint32_t)(((uintptr_t)saddr) >> 12))
+#define	ECCAREACR(ecc, saddr) \
+	(((uint32_t)ecc << 31) | (uint32_t)(((uintptr_t)saddr) >> 12))
+
+#ifndef ARRAY_SIZE
+#define	ARRAY_SIZE(a)		(sizeof(a) / sizeof(a[0]))
+#endif
+
+static const struct rzg2_ecc_conf {
+	uint32_t fusaareacr;
+	uint32_t eccareacr;
+} rzg2_ek874_conf[] = {
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+};
+
+/* Setup DRAM ECC configuration registers */
+void bl2_enable_ecc_ek874(void)
+{
+	const struct rzg2_ecc_conf *conf;
+	int n;
+	int nb_of_conf;
+
+	conf = rzg2_ek874_conf;
+	nb_of_conf = ARRAY_SIZE(rzg2_ek874_conf);
+
+	for (n = 0; n < nb_of_conf; n++, conf++) {
+		mmio_write_32((uintptr_t)((uint32_t *)DFUSAAREACR + n), conf->fusaareacr);
+		mmio_write_32((uintptr_t)((uint32_t *)DECCAREACR + n), conf->eccareacr);
+	}
+
+}
diff --git a/plat/renesas/rcar/bl2_rcar_setup.c b/plat/renesas/rcar/bl2_rcar_setup.c
index ed6f9da..a18b105 100644
--- a/plat/renesas/rcar/bl2_rcar_setup.c
+++ b/plat/renesas/rcar/bl2_rcar_setup.c
@@ -697,6 +697,13 @@ static void rcar_bl2_early_platform_setup(const meminfo_t *mem_layout)
 		qos_init();
 	}
 
+#if (RZG_DRAM_EK874_ECC == 1)
+	{
+		extern void bl2_enable_ecc_ek874(void);
+		bl2_enable_ecc_ek874();
+	}
+#endif
+
 	if ((modemr_boot_dev == MODEMR_BOOT_DEV_EMMC_25X1) ||
 	    (modemr_boot_dev == MODEMR_BOOT_DEV_EMMC_50X8)) {
 		/* Initialize eMMC */
diff --git a/plat/renesas/rcar/platform.mk b/plat/renesas/rcar/platform.mk
index 275b08a..25d7e7e 100644
--- a/plat/renesas/rcar/platform.mk
+++ b/plat/renesas/rcar/platform.mk
@@ -33,6 +33,7 @@ BL2_SOURCES		+=	${RCAR_GIC_SOURCES}				\
 				plat/renesas/rcar/drivers/error/bl2_int_error.c	\
 				plat/renesas/rcar/aarch64/rcar_helpers.S	\
 				plat/renesas/rcar/bl2_rcar_setup.c		\
+				plat/renesas/rcar/bl2_fusa.c			\
 				plat/renesas/rcar/aarch64/rcar_common.c		\
 				plat/renesas/rcar/aarch64/rcar_drivers_common.c	\
 				plat/renesas/rcar/drivers/io/io_rcar.c		\
@@ -366,6 +367,12 @@ RCAR_DRAM_DDR3L_MEMDUAL :=1
 endif
 $(eval $(call add_define,RCAR_DRAM_DDR3L_MEMDUAL))
 
+# Process RZG_DRAM_EK874_ECC flag
+ifndef RZG_DRAM_EK874_ECC
+RZG_DRAM_EK874_ECC :=0
+endif
+$(eval $(call add_define,RZG_DRAM_EK874_ECC))
+
 # Process RCAR_BL33_ARG0 flag
 ifdef RCAR_BL33_ARG0
 $(eval $(call add_define,RCAR_BL33_ARG0))
-- 
2.7.4

